### DISCUSSION TITLE
A torus/donut collision shape would be helpful.

### DISCUSSION THREAD

**Discussion by Melvin8D** [godotengine/godot-proposals#6244] [3D]
Torus shapes are a common shape for shockwave collisions in games. I was wanting to put a torus collision shape that can dynamically expand. I mainly wanted a 3d shape, but I've seen people on reddit wanting a 2d one as well.
Trying to implement with basic godot editor shenanigans is kinda tricky, and I have no knowledge of low level engine stuff. 
Using a method involving 2 cylinder collision shapes where if your only touching the outer one wouldn't be the best, as typically you'd want to define the outside radius as a set amount, however depending on how wide a collider is it would sometimes always be "touching" the preventor even when in the damager.

**Comment by venilark** [comment]
As a workaround until they implement this, you can use a CSGTorus, give it a unique collision mask and for your character, use an Area that only scans that layer. Wouldn't be necessary if the CSG shapes had a body_entered method but right now I think it is the only way.

**Reply by aXu-AP** [reply]
Second workaround is to add MeshInstance with TorusMesh, then in editor select Mesh -> Create TriMesh Collision Sibling, and finally hide/delete the MeshInstance. Minus side is that tweaking the results requires extra step.

**Reply by Melvin8D** [reply]
I did do a semi-work around, where I used 16 box meshes to sort of mimic a torus shape. This allowed me to just do a bit of math and get something close to an expanding torus. Not exact but a good enough approximation. I originally made this post because I was just completely stumped in any way shape or form to find a way to do it.

**Reply by venilark** [reply]
@aXu-AP for static rings that's a good solution but with the torus csg you can easily change the inner and outer radius and make it expand over time easily, with a fixed collision trimesh no idea how that would work

**Comment by venilark** [comment]
has this been considered?

**Reply by Calinou** [reply]
Unlike most primitive collision shapes, a torus is not convex, which means I'm not sure if it can be supported in a fast manner.

Also, consider that GodotPhysics already has trouble with cylinder collision shapes, and that [Jolt](https://github.com/jrouwe/JoltPhysics) doesn't support torus collision shapes.

**Reply by venilark** [reply]
looks like everybody is eyeing on Jolt which is good news
in this case the CSGTorus trick seems to be the only way although I don't like it at all
